### PR TITLE
Better cudagraph enablement for dynamic_shapes=True with static shapes

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101813

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/compile_fx.py
Comment (ezyang) [inline torch/_inductor/compile_fx.py:311]: Instead of this use free_symbols from symbolic_shapes

**Review by ezyang — APPROVED** [review]
Please check comments

File: torch/_inductor/compile_fx.py
Comment (voznesenskym) [inline torch/_inductor/compile_fx.py:303]: @ezyang said "Stop testing sympy.Integer on size lol" because he has told me this before, but I am dense.

**Comment by ezyang** [comment]
Also when we get nonzero in graph we will have to add a more sophisticated test here

**Comment by vadimkantorov** [comment]
is it possible to use cuda graphs in practice if the shapes are not static but e.g. a small collection of possible shapes?

**Review by eellison — COMMENTED** [review]
Few things i'm worried about :

1 - 

First iteration cudagraphs
Second iteration fallback to dynamic 

This will split the caching allocator in 2 and likely OOM. We could make the eager allocator "absorb" the cudagraph allocator and invalidate all the cudagraphs but this is not implemented.


2 - 

How are preserving the use case where we re-cudagraphs on new shapes ? 

3 - 

Grepping around, there are users that just set `cudagraphs=True`. I'm worried how we preserve the behavior there. Should setting `cudagraphs=True` turn off assume dynamic ?

Can we resolve these before landing ? cc @ngimel

File: torch/_inductor/compile_fx.py
Comment (voznesenskym) [reply torch/_inductor/compile_fx.py:301]: Ehh? is_symbolic takes 2 values, and the algos are the same. Not worth.

**Comment by voznesenskym** [comment]
> Few things i'm worried about :
> 
> 1 -
> 
> First iteration cudagraphs Second iteration fallback to dynamic
> 
> This will split the caching allocator in 2 and likely OOM. We could make the eager allocator "absorb" the cudagraph allocator and invalidate all the cudagraphs but this is not implemented.

This case is handled in the PR - this can only happen if we are in automatic dynamic.

> 2 -
> 
> How are preserving the use case where we re-cudagraphs on new shapes ?

Can you elaborate?

> 3 -
> 
> Grepping around, there are users that just set `cudagraphs=True`. I'm worried how we preserve the behavior there. Should setting `cudagraphs=True` turn off assume dynamic ?
> 
> Can we resolve these before landing ? cc @ngimel

Can you elaborate the concern here that is not covered by (1)? If someone set cudagraphs=True and dynamic_shapes=True, 

The matrix is as follows:

|         | automatic_dynamic=True | automatic_dynamic=False |
| ----------- | ----------- | ----------- |
| assume_static_by_default=True      | `cudagraphs=False` (explicit in this PR)     |  `cudagraphs=True  if no mark_dynamic, else False`
| assume_static_by_default=False   | `cudagraphs=False` (inputs will be dynamic)        | `cudagraphs=False` (inputs will be dynamic)

**Comment by eellison** [comment]
> 2 -

How are preserving the use case where we re-cudagraphs on new shapes ?

The following case:

```
@torch.compile(mode='reduce-overhead')
def run_model(model, inp):
    return model(inp)

# run with one batch size
run_model(resnet, torch.rand([1, 3, 255, 255], device="cuda")
# run again with different batch size
run_model(resnet, torch.rand([16, 3, 255, 255], device="cuda")
```

For 3 - 

Thanks for the matrix. Could you clarify what is the proposed end default setting for both `assume_static_by_default` and `automatic_dynamic` ?

**Comment by voznesenskym** [comment]
Sorry @eellison fat fingered and edited your comment somehow.

It used to say:

2 -

How are preserving the use case where we re-cudagraphs on new shapes ?

The following case:

@torch.compile(mode='reduce-overhead')
def run_model(model, inp):
    return model(inp)

# run with one batch size
run_model(resnet, torch.rand([1, 3, 255, 255], device="cuda")
# run again with different batch size
run_model(resnet, torch.rand([16, 3, 255, 255], device="cuda")
For 3 -

Thanks for the matrix. Could you clarify what is the proposed end default setting for both assume_static_by_default and automatic_dynamic ?

**Comment by voznesenskym** [comment]
@eellison 

for 2: Which regime are you asking about?
For 3: True/True

**Comment by voznesenskym** [comment]
> is it possible to use cuda graphs in practice if the shapes are not static but e.g. a small collection of possible shapes? if yes, it would be good to have some monitorability / logging options to understand how caching / recompilation / graph recapture would work
> 
> a small collection of possible shapes is more practical regime for training time

We kinda discussed sets of hinted values to do better optimization than the current regime of static vs dynamic.

However, we are a few steps away from that - the current set of changes we need to make is moving from checking config to introspecting types, as this PR does.

**Comment by voznesenskym** [comment]
image

Perf identical

**Comment by eellison** [comment]
@voznesenskym we would expect the perf to be the same because we dont run different input shapes in training for cudagraphs. maybe we can VC

File: torch/_inductor/compile_fx.py
Comment (ezyang) [reply torch/_inductor/compile_fx.py:311]: No, I mean, use this:

```
def free_symbols(val: Union[SymInt, torch.Tensor]) -> Set[sympy.Symbol]:
    if isinstance(val, (SymInt, SymFloat)):
        return val.node.expr.free_symbols
    elif isinstance(val, (int, float, bool)):
        return set()
    elif isinstance(val, torch.Tensor):
        r = set()
        for s in val.size():
            r |= free_symbols(s)
        for s in val.stride():
            r |= free_symbols(s)
        r |= free_symbols(val.storage_offset())
        return r
    else:
        raise AssertionError(f"cannot compute free_symbols of {val}")
```

in torch.fx.experimental.symbolic_shapes

**Comment by eellison** [comment]
For 2: whatever we are proposing as the default.
```
@torch.compile(mode='reduce-overhead')
def run_model(model, inp):
    return model(inp)

# run with one batch size
run_model(resnet, torch.rand([1, 3, 255, 255], device="cuda")
# run again with different batch size
run_model(resnet, torch.rand([16, 3, 255, 255], device="cuda")
``` 

should still work and cudagraph both invocations

**Comment by voznesenskym** [comment]
> @voznesenskym we would expect the perf to be the same because we dont run different input shapes in training for cudagraphs. maybe we can VC

No. It's because automatic_dynamic=True is the default, and so we still don't use cudagraphs for dynamic.

> For 2: whatever we are proposing as the default.
> 
> ```
> @torch.compile(mode='reduce-overhead')
> def run_model(model, inp):
>     return model(inp)
> 
> # run with one batch size
> run_model(resnet, torch.rand([1, 3, 255, 255], device="cuda")
> # run again with different batch size
> run_model(resnet, torch.rand([16, 3, 255, 255], device="cuda")
> ```
> 
> should still work and cudagraph both invocations

I think it still would? This would be a problem for enabling automatic_dynamic, not for enabling dynamic_shapes, right?

With dynamic_shapes=True + assume_static=True + automatic_dynamic=True (the current setup) we don't enable cudagraphs because of automatic_dynamic.

With dynamic_shapes=True + assume_static=True + automatic_dynamic=False, your code snippet is safe because that would essentially compile 2 static graphs as if we had dynamic_shapes=False.

Does that make sense?

File: torch/_inductor/compile_fx.py
Comment (voznesenskym) [reply torch/_inductor/compile_fx.py:311]: Oh very nice.

File: torch/_inductor/compile_fx.py
Comment (ezyang) [inline torch/_inductor/compile_fx.py:307]: nit: Don't test booleans with `is True`

**Comment by ezyang** [comment]
@voznesenskym, it sounds like you want to land this PR as is and deal with the non-dynamic to dynamic transition in a separate PR. But you can't land this as is because `automatic_dynamic_shapes` defaults to True, so by switching the CUDA graphs test to `has_dynamism` you have essentially unconditionally turned off CUDA graphs entirely, UNLESS someone manually changes that config.

I have a different proposed interim configuration: if mode="reduce-overhead", we DISABLE dynamic shapes entirely and do recompiles (because the whole point is you wanted cuda graphs.) This still gets us in trouble if Elias wants cuda graphs by default, but we can cross that bridge when we get there.

**Review by ezyang — CHANGES_REQUESTED** [review]
New changes are not ok

**Comment by voznesenskym** [comment]
> @voznesenskym, it sounds like you want to land this PR as is and deal with the non-dynamic to dynamic transition in a separate PR. But you can't land this as is because `automatic_dynamic_shapes` defaults to True, so by switching the CUDA graphs test to `has_dynamism` you have essentially unconditionally turned off CUDA graphs entirely, UNLESS someone manually changes that config.
> 
> I have a different proposed interim configuration: if mode="reduce-overhead", we DISABLE dynamic shapes entirely (this is a harder disable than automatic dynamic, because even if we detected a dim is dynamic we still want to recompile) and do recompiles (because the whole point is you wanted cuda graphs.) This still gets us in trouble if Elias wants cuda graphs by default, but we can cross that bridge when we get there.

Well, https://github.com/pytorch/pytorch/pull/100815 defaults it to False, and the idea was to land this after that.

I apologize though, I was not sufficiently clear. It also puts us at an impasse with the PR above. 

I like your solution for mode=reduce-overhead more though. Let me give that a spin.

File: torch/__init__.py
Comment (eellison) [inline torch/__init__.py:1505]: Should we also do this on setattr of `cudagraphs` ? many users will just manipulate the config themselves instead of passing this mode in

Comment (voznesenskym) [reply torch/__init__.py:1505]: Ed's suggestion was based on mode, if we do this, then we might as well leave it as it is entirely right?

File: torch/_inductor/compile_fx.py
Comment (ezyang) [inline torch/_inductor/compile_fx.py:306]: comment is stale now

File: torch/__init__.py
Comment (ezyang) [reply torch/__init__.py:1505]: Actually I like Elias's suggestion. Essentially, the change we should make is that when we consult automatic_dynamic, we should also consult triton.cudagraphs. If triton.cudagraphs is true, this forces automatic_dynamic to False (in fact, it is a little stronger than that: if triton.cudagraphs is true, we must never make things dynamic, even if the user asked us to do so!)

Comment (ezyang) [reply torch/__init__.py:1505]: It's a bit annoying that we can't delete this config option anymore but it seems exactly what you want to trigger the cudagraphs behavior.

**Pull request closed without merging**
